random/seeded.ts

```diff
@@ -0,0 +1,220 @@
+    // equivalently to Math.random()
+    this.random = this.random.bind(this);
+  }
+
+  /**
+   * The state for the random number generator.
+   * @returns The state as a 2-tuple.
+   *
+   * @example Usage
+   * ```ts
+   * import { SeededPrng } from "@std/random";
+   * import { assertEquals } from "@std/assert";
+   *
+   * const prng = new SeededPrng({ seed: 14614327452668470620n });
+   * assertEquals(prng.state, [13062938915293834817n, 10846994826184652623n]);
+   * ```
+   */
+  get state(): Readonly<State> {
+    return [this.#state, this.#inc];
+  }
+
+  set state(value: Readonly<State>) {
+    this.#state = value[0] & U64_MAX;
+    this.#inc = (value[1] | 1n) & U64_MAX;
+  }
+
+  /**
+   * Generates a random number using the current state of the prng.
+   * @returns A random number between 0 and 1
+   *
+   * @example Usage
+   * ```ts
+   * import { SeededPrng } from "@std/random";
+   * import { assertEquals } from "@std/assert";
+   *
+   * const prng = new SeededPrng({ seed: 3377567226465013078n });
+   *
+   * assertEquals(prng.random(), 0.34011089405976236);
+   * assertEquals(prng.random(), 0.6603851807303727);
+   * assertEquals(prng.random(), 0.4863424440845847);
+   * ```
+   */
+  random(): number {
+    return uint32ToFloat64(this.#randomUint32());
+  }
+
+  /**
+   * Generates a new seed from the current prng that can be used to create a new prng.
+   * @returns The new seed
+   *
+   * @example Usage
+   * ```ts no-assert
+   * import { SeededPrng } from "@std/random";
+   * import { assertEquals } from "@std/assert";
+   *
+   * const prng = new SeededPrng({ seed: 14614327452668470620n });
+   * assertEquals(prng.randomSeed(), 7382748496997062591n);
+   * ```
+   */
+  randomSeed(): bigint {
+    return this.#randomUint64();
+  }
+
+  // port of minimal version at https://www.pcg-random.org/download.html
+  #randomUint32(): bigint {
+    const oldState = this.#state;
+
+    this.#state = (this.#state * 6364136223846793005n + (this.#inc | 1n)) %
+      U64_CEIL;
+
+    const xorShifted = (((oldState >> 18n) ^ oldState) >> 27n) % U32_CEIL;
+    const rot = (oldState >> 59n) % U32_CEIL;
+
+    return ((xorShifted >> rot) | (xorShifted << ((-rot) & 31n))) % U32_CEIL;
+  }
+
+  #randomUint64(): bigint {
+    const hi = this.#randomUint32();
+    const lo = this.#randomUint32();
+    return (hi << 32n) | lo;
+  }
+
+  /**
+   * Get a random state tuple suitable for usage by `SeededPrng`.
+   */
+  #randomState(): State {
+    return [
+      this.#randomUint64(),
+      this.#randomUint64(),
+    ];
+  }
+
+  /**
+   * Convert a user-provided scalar seed value to a state tuple suitable for
+   * usage by `SeededPrng`.
+   *
+   * Aims to provide good results for both well-randomized seeds (such as
+   * 64-bit outputs from `crypto.getRandomValues`) and low-entropy seeds (such
+   * as 1, 2, 3, keyboard mashing, etc.).
+   */
+  static #seedToState(seed: bigint): State {
```

Comment:
Couldn't that be just a separate function? I think a static private method is probably not very useful.

Reply:
Yeah probably. I'm still not settled on how best to handle seeding tbh — upon looking into it further it seems like the C/C++ reference implementation have a good way of initializing via seed data, but that seed consists of 2 parts (`initstate` and `initseq`), both of which are uint64. So just accepting 16 bytes of state as a seed could work, even though it's no longer a scalar value (unless you conceptualize it as a single 128-bit int, but JS doesn't support 128-bit ints via any web APIs etc)